tools/cndump/kvset_dump.c

```diff
@@ -0,0 +1,398 @@
+ */
+
+#include <assert.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#include <hse/error/merr.h>
+#include <hse_ikvdb/diag_kvdb.h>
+#include <hse_ikvdb/omf_kmd.h>
+#include <hse_util/parse_num.h>
+#include <hse_util/fmt.h>
+
+#include <mpool/mpool_structs.h>
+#include <mpool/mpool.h>
+
+#include <cn/omf.h>
+
+#include "cndb_reader.h"
+#include "cndb_record.h"
+#include "fatal.h"
+#include "globals.h"
+
+struct opt {
+    uint64_t kvset_id;
+};
+
+static struct opt opt;
+
+struct dump_mblock {
+    struct mblock_props props;
+    void               *data;
+};
+
+struct dump_kvset {
+    struct cndb_rec     rec;
+    struct dump_mblock  hblk;
+    struct dump_mblock *kblkv;
+    struct dump_mblock *vblkv;
+};
+
+static void
+help(void)
+{
+    printf("Usage: %s kvset [options] <kvdb_home> <kvset_id>\n", progname);
+    printf("Pretty print a kvset.\n"
+        "  <kvdb_home>   KVDB home directory.\n"
+        "  <kvset_id>    kvset ID (from the CNDB log).\n"
+        "Options:\n"
+        "  -h    print help\n"
+        "Notes:\n"
+        "  - Kvset IDs are global to the KVDB, so there is no need to specify the KVS.\n");
+}
+
+unsigned int
+kvset_cmd_parse_args(int argc, char **argv)
+{
+    unsigned int consumed;
+    const char *arg;
+    int c;
+
+    /* argv[0] should hold command name */
+    assert(argc > 0);
+    consumed = 1;
+    optind = 0;
+
+    while ((c = getopt(argc, argv, "+:h")) != -1) {
+        consumed++;
+        switch (c) {
+        case 'h':
+            help();
+            exit(0);
+        default:
+            syntax("invalid kvset option -%c", c);
+            break;
+        }
+    }
+
+    if (consumed == argc)
+        syntax("missing <kvdb_home> parameter");
+
+    kvdb_home = argv[consumed++];
+
+    if (consumed == argc)
+        syntax("missing <kvset_id> parameter");
+
+    arg = argv[consumed++];
+    if (parse_u64(arg, &opt.kvset_id))
+        syntax("invalid kvset ID: '%s' (cannot convert to uint64_t)", arg);
+
+    if (consumed != argc)
+        syntax("unexpected parameter: '%s'", argv[consumed]);
+
+    return consumed;
+}
+
+/* fmt_data - format data as ascii or hex
+ *
+ * use two static buffers, allows 2 simultaneous uses per printf
+ * and buffer management is hidden from users
+ *
+ * NOT THREAD SAFE!
+ */
+char *
+fmt_data(void **buf_ptr, size_t *bufsz_ptr, const void *data, size_t data_len, size_t limit)
+{
+    char *buf = *buf_ptr;
+    size_t bufsz = *bufsz_ptr;
+    size_t len = (limit && limit < data_len) ? limit : data_len;
+    size_t output_len = fmt_hex_max_output_len(len);
+
+    if (bufsz < output_len) {
+        void *p = realloc(buf, output_len);
+        if (!p)
+            fatal("realloc", merr(ENOMEM));
+        buf = *buf_ptr = p;
+        bufsz = *bufsz_ptr = output_len;
+    }
+
+    fmt_hex(buf, bufsz, data, len);
+
+    return buf;
+}
+
+#if 0
+char *
+fmt_key(const void *v, int off, int len)
```

Comment:
what is the story with this?

Reply:
It's part of `cn_kbdump`. Will delete it until/if we update `cndump` to do more of what `cn_kbdump` does.

Reply:
removed.